### DISCUSSION TITLE
Removal of user keymaps from QMK

### DISCUSSION THREAD

**Discussion by manna-harbour** [manna-harbour/miryoku#287] [Announcements]
QMK have introduced a breaking change that removes user keymaps, including Miryoku QMK. See [the announcement](https://github.com/qmk/qmk_firmware/blob/master/docs/ChangeLog/20231126.md#external-userspace-22222) for details.

As a result, Miyoku QMK is not longer present in QMK master, and references to QMK master in the Miryoku QMK documentation are now out of date. In particular, Miryoku QMK can't be [built directly from QMK master](https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#qmk-master), and the `miryoku` branch can no longer be [merged with QMK master](https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#merge). Miryoku QMK can still be [built directly from the `miryoku` branch](https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#miryoku-qmk-development-branch), where a more recent version of QMK is not required.

As a temporary work around, the last commit still containing user keymaps in QMK can be used as a substitute for QMK master. That commit is at tag `user-keymaps-still-present`, and has been made available as a branch for merging at https://github.com/manna-harbour/qmk_firmware/tree/user-keymaps-still-present. That branch can be used in place of QMK master in the Miryoku QMK documentation to build from directly, or merge with Miryoku QMK, if a more recent version of QMK than is present in the `miryoku` branch is required. For merging using workflow builds, use `manna-harbour/qmk_firmware/user-keymaps-still-present` with the workflow merge option.

To update for changes introduced to QMK after that commit, such as new keyboards, files will need to be copied manually between branches. To add Miryoku QMK to QMK, copy the contents of https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku plus the required keyboard (e.g. https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/keyboards/a_dux/keymaps/manna-harbour_miryoku) or community layout (e.g. https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/layouts/community/split_3x5_3/manna-harbour_miryoku) keymaps from your Miryoku QMK clone to the corresponding locations in your QMK clone. To add individual keyboard support from QMK to Miryoku QMK, copy the required keyboard directory from your QMK clone to the corresponding directory in your Miryoku QMK clone.

In future, Miryoku QMK will be updated to use the new [user keymap overlay feature "External QMK Userspace"](https://docs.qmk.fm/#/newbs_external_userspace?id=external-qmk-userspace) instead of being a QMK fork.

@drashna has made a copy of Miryoku QMK adapted for this feature as a proof-of-concept, available at https://github.com/drashna/qmk_userspace/tree/miryoku. This can be used now for testing, but won't be updated and isn't supported.

**Comment by rvannoord** [comment]
Has Miryoku QMK already been updated to use the new user keymap overlay feature "External QMK Userspace"?
I've been trying to find answers to this in your repo but no luck (probably my lack of understanding).

**Reply by manna-harbour** [reply]
It will be announced here when it's updated, so subscribe to this discussion for notifications. For now, see the work arounds above.

**Reply by peterjc** [reply]
The main README needs updating, and perhaps a link to this discussion. e.g. https://github.com/manna-harbour/miryoku_qmk/pull/43

**Comment by sudosteve** [comment]
I'm interested in this feature. My keyboard is supported in the main qmk repo but hasn't been added here yet. I was considering submitting a PR but it seems the right way to get this support is by migrating to the new user_space way of doing things.

What work is required to push this forward?

**Reply by manna-harbour** [reply]
> I'm interested in this feature. My keyboard is supported in the main qmk repo but hasn't been added here yet. I was considering submitting a PR but it seems the right way to get this support is by migrating to the new user_space way of doing things.

Yes, keyboard definitions won't be added to the `miryoku` branch. For now, see the work arounds above.

> What work is required to push this forward?

A new repo needs to be created based on the "userspace" feature. See above for a proof-of-concept example that you could try using now. Ideally the history of the various historical `miryoku` feature branches would be included, but that might end up in a separate repo. Then the custom workflows need to be adapted.

**Reply by peterjc** [reply]
In https://github.com/peterjc/qmk_userspace/pull/4 I was able to add the `/users/miryoku/` folder and `layouts/*/miryoku/*` to my `qmk_userspace` repository (cherry-picking from drashna's proof-of-concept at https://github.com/drashna/qmk_userspace/tree/miryoku for this), and compile for my new keyboard with current QMK using GitHub Actions 🚀 (still waiting on the PCB etc to test this). Thank you!

Would an official "qmk_miryoku_userspace" repository also need the `keyboard/*/keymaps/manna-harbour_miryoku/*` files declaring miryoku layouts?

eg https://github.com/qmk/qmk_firmware/tree/master/keyboards/3w6/ only has the default mapping and does not have the Miryoku files at https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/keyboards/3w6/keymaps/manna-harbour_miryoku

**Reply by manna-harbour** [reply]
@peterjc 
> Would an official "qmk_miryoku_userspace" repository also need the `keyboard/*/keymaps/manna-harbour_miryoku/*` files declaring miryoku layouts?

It should have `/users/manna-harbour_miryoku/...`, `layouts/.../manna-harbour_miryoku/...`, and `keyboards/.../keymaps/manna-harbour_miryoku/...`. Where `manna-harbour_miryoku` has been renamed to `miryoku` in Drashna's repo, and probably will also be renamed in the new repo (since the `manna-harbour_miryoku` name was required due to QMKs naming rules which will no longer apply). `keyboards/.../keymaps/...` is the usual keymap location and is required for most keyboards. `layouts/...` is only for keyboards that support a community layout.

**Reply by peterjc** [reply]
Got it. I will give that a try...

I was thinking you might drop (all? most?) `keyboards/.../keymaps/manna-harbour_miryoku/...` files as part of this move from being a QMK fork, to a QMK userspace based repository. When or if anyone misses an older keyboard, ask them to help with testing a QMK update to add the appropriate community layout to the keyboard. I was asking about that on the QMK Discord today, and that kind of "housekeeping" really does still need testing on an actual keyboard (there were a lot of pull requests like that marked stale for a lack of testing).

**Reply by peterjc** [reply]
I was surprised that there were only 41 commits left on the `miryoku` branch after this:

```console
git-filter-repo \
    --path "users/manna-harbour_miryoku/" \
    --path-regex "layouts/.*/manna-harbour_miryoku/" \
    --path-regex "keyboards/.*/keymaps/manna-harbour_miryoku/" \
    --filename-callback "return filename.replace(b'/manna-harbour_miryoku/', b'/miryoku/')"
```

(Using `--path-regex` not `--path-glob` as can be more than one directory nesting with keyboard variants.)

The first commit is then from June 2019, "[Keymap] miryoku layout (#6171)" from https://github.com/manna-harbour/miryoku_qmk/commit/4be1dfcee48daeb77e966b08ad9ca51e02253acf - so I assume this is complete?

Finally, a little more git magic is still needed to transplant that 41 commit branch onto a fork of https://github.com/qmk/qmk_userspace - and then some polish like the main README and GitHub Actions setup (see earlier work by drashna).

**Reply by peterjc** [reply]
I got a lot further, but then GitHub Actions hit `include users/manna-harbour_miryoku/post_rules.mk` lines breaking due to the folder renaming.

I think it is simplest (and more historically accurate) to delay the renaming and search-and-replace fix to after the git history pruning.

**Reply by peterjc** [reply]
Looking good now with GitHub Actions - seven with warnings, four errors:

* `atreus/feather:miryoku` - The firmware is too large! 30184/28672 (1512 bytes over)
* `for_science:miryoku` macro expansion issue
* `ily58/r2g:miryoku` - The firmware is too large! 29040/28672 (368 bytes over)
* `waterfowl:miryoku` - The firmware is too large! 29284/28672 (612 bytes over)

This is currently a private repo, with @manna-harbour invited as a collaborator. The branch could be pulled into an official repo, or even transfer ownership - whatever works for you.

**Reply by manna-harbour** [reply]
@peterjc 
> I was thinking you might drop (all? most?) `keyboards/.../keymaps/manna-harbour_miryoku/...` files as part of this move from being a QMK fork, to a QMK userspace based repository. 

It's not really necessary to create a community layout for every keyboard. If many keyboards share the same layout then a new community layout should be created, and if a keyboard has the same layout as an existing community layout then it should be converted. But they are all changes in QMK. For Miryoku QMK, the existing keyboards and layouts keymaps can be migrated as-is.

**Reply by peterjc** [reply]
Sure.

I meant if/when older keyboards like 3w6 were updated in https://github.com/qmk/qmk_firmware/blob/develop/keyboards/3w6/keymaps/default/keymap.c to use LAYOUT_split_3x5_3 etc for the split_3x5_3 community layout, then you could drop https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/keyboards/3w6/keymaps/manna-harbour_miryoku

**Reply by manna-harbour** [reply]
@peterjc Yes, absolutely! And that has been done previously.

**Reply by peterjc** [reply]
Is the proof of principle QMK user space with Miryoku history useful to you? Should I make it public for comment?

**Reply by manna-harbour** [reply]
@peterjc I'm planning on making a repo that includes the full history from all of the original branches i.e. before squashing for QMK PRs. When I get to that point I'll compare with your repo, including the commits you've added on top, so thanks for making it! At the moment I'm working on a new build system, but after that I'll work on the history.

**Reply by peterjc** [reply]
That sounds good. There is a hack to compile via the QMK user space  repository GitHub Actions even with keyboard specific C code - eg https://github.com/peterjc/qmk_userspace/pull/4 - I need to finish soldering my keyboard to test this for real 😉

**Reply by peterjc** [reply]
I have now tested the firmware from https://github.com/peterjc/qmk_userspace/pull/4 on my own keyboard with `LAYOUT_split_3x5_3`, demonstating the GitHub Actions approach via QMK Userspace. I look forward to the official QMK Miryoku Userspace 😄

**Reply by manna-harbour** [reply]
@peterjc
I didn't notice any keyboard specific code in the linked PR? But the new build system will incorporate [the outboards feature from Miryoku ZMK](https://github.com/manna-harbour/miryoku_zmk#outboards), so that could be used for keyboard specific code, and also to build keyboards that maintain their keyboard definitions outside QMK.

**Reply by peterjc** [reply]
That repo has my own 36-key keyboard “tutte_coxeter_36k” QMK firmware supporting the community split 3x5_3 layout. The keyboard has has a default layout, vial, and on that branch is also built with Miryoku.

I have since also started to look at how ZMK handles out-of-tree keyboards.

**Comment by winmillwill** [comment]
Just a note for some issues I encountered: (EDIT: was trying to compile for a newer board added after userspace removal)
 - needed to apply this fix to drashna's userspace example: https://docs.qmk.fm/ChangeLog/20240825#key-override-keymap-c-signature-change-24120
 - could not just copy new keyboard files to pre-removal qmk because it locates by looking for rules.mk and my board doesn't have that file and generating it using the hidden qmk command left some data structure unpopulated and made the qmk cli error out

What ultimately let me compile was following the hardware support steps here https://github.com/manna-harbour/miryoku/discussions/85 in drashna's miryoku branch  (not tag, checking out the branch is annoying but doable with `git switch -c miryoku --track origin/miryoku`) together with the `key_overrides` change above.

**Reply by peterjc** [reply]
That fix is in drashna's proof-of-principle qmk userspace miryoku branch now: https://github.com/drashna/qmk_userspace/commit/55dda9d792dc491f6c6f26e58208c93d2d213af2

**Comment by manna-harbour** [comment]
To build with current QMK also see https://github.com/manna-harbour/miryoku/discussions/361.

**Reply by rvannoord** [reply]
Cool! So does that mean we won’t really need the `user-keymaps-still-present` branch anymore?

**Reply by manna-harbour** [reply]
@rvannoord This is only if you're copying Miryoku into current QMK master.

**Comment by maatthc** [comment]
Using the instructions I've found here, I've created a [userspace](https://github.com/maatthc/qmk_userspace) for **beekeeb/piantor** using the latest version of QMK (I needed that to give [Predictive Tap Hold](https://github.com/jgandert/qmk_modules/tree/main/predictive_tap_hold) a try). 

Two additional changes I had to do:

- Rename the keymap : [Keymap names must contain only a-z, 0-9 and _!](https://github.com/maatthc/qmk_userspace/commit/5da8f42fe31f5d9b282621b5e559f68a442c1ef1)
- [Update miryoku mouse keycodes](https://github.com/maatthc/qmk_userspace/commit/afcb621aeaa19ec02bc0e5ab57bee7c9691ed77e)

**Comment by riderkirk** [comment]
I am new to this whole keyboard thing. I am trying to get setup using the instructions here to flash miryoku to my crkbd v4.1.But first I just wanted to add miryoku_qmk to qmk and test it by compiling rev1. I can compile the firmware for rev1 when I am in miryoku_qmk. But in my own branch where I tried to add miryoku_qmk to the most recent version of qmk I can't get that working. I copied https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku as stated above. But I didn't find any manna-harbour_miryoku keymap for crkbd to copy over. Which raised the question, how is rev1 able to compile in miryoku_qmk if that keymap doesn't exist there? Again, I am new to this keyboard firmware stuff so forgive me if my confusion is ignorant. Am I even on the right track for getting this working for corne v4.1?

**Reply by peterjc** [reply]
@riderkirk It is tricky, and will remain so until Miryoku QMK is updated for the QMK upheaval. If using an older QMK is fine for you (before all the keyboard removal stuff which Miryoku still needs), try the advice on https://github.com/manna-harbour/miryoku_qmk/pull/40 perhaps?

**Reply by argbet21** [reply]
@riderkirk Hey, I'm using Corne v4.1 too (`crkbd/rev4_1/standard` more precisely) -- did you manage to flash it? Would you mind sharing how you managed to flash Miryoku on it?

**Reply by yehorb** [reply]
I was able to [make Miryoku work on `crkbd/rev4_1/standard`](https://github.com/manna-harbour/miryoku/discussions/287#discussioncomment-17079420).

**Comment by yehorb** [comment]
I spent the whole day untangling QMK build system and setting Miryoku up to work with the latest QMK master. Thankfully, I was able to fix and build it for `crkbd/rev4_1/standard`.

Here is [my userspace](https://github.com/yehorb/qmk_userspace/tree/miryoku), here is the explanation of [how to set it up](https://docs.qmk.fm/newbs_external_userspace). The tag I linked is my first successful build of Miryoku with default options, you can check later commits for customization, or follow the Miryoku docs. The only minor inconvenience is that my keyboard layout is `split_3x6_3_ex2`, and those 2 extra keys may be redundant. Copy the `keyboards/crkbd/rev4_1/standard/keymaps/miryoku_split_3x6_3_ex2` directory to a new directory corresponding to your layout, for example `keyboards/crkbd/rev4_1/standard/keymaps/miryoku_split_3x6_3`, and edit the `config.h` to use  the correct macro. Then Miryoku should build and flash for your keyboard no problem.

Build options and further [Miryoku Customisation](https://github.com/manna-harbour/miryoku/discussions/85) work as expected.

Regarding the whole "porting" process. Dear @manna-harbour, I am willing to help to set up an updated, "userspace" based Miryoku. I think I have a good idea of what has to be done and how can it be automated. If you are willing to move forward with it let's discuss the scope (what exactly needs to be ported) and I will do my best to update the code and build instructions.
